pkg/registry/apps/daemonset/storage/storage.go

```diff
@@ -108,3 +136,67 @@ func (r *StatusREST) GetResetFields() map[fieldpath.APIVersion]*fieldpath.Set {
 func (r *StatusREST) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
 	return r.store.ConvertToTable(ctx, object, tableOptions)
 }
+
+// ScaleREST implements a Scale for DaemonSet.
+type ScaleREST struct {
+	store *genericregistry.Store
+}
+
+// ScaleREST implements only Getter, /scale is read-only by design
+var _ = rest.Getter(&ScaleREST{})
+var _ = rest.GroupVersionKindProvider(&ScaleREST{})
+
+// GroupVersionKind returns GroupVersionKind for DaemonSet Scale object
+func (r *ScaleREST) GroupVersionKind(containingGV schema.GroupVersion) schema.GroupVersionKind {
+	switch containingGV {
+	case appsv1beta1.SchemeGroupVersion:
+		return appsv1beta1.SchemeGroupVersion.WithKind("Scale")
+	case appsv1beta2.SchemeGroupVersion:
+		return appsv1beta2.SchemeGroupVersion.WithKind("Scale")
+	default:
+		return autoscalingv1.SchemeGroupVersion.WithKind("Scale")
+	}
+}
+
+// New creates a new Scale object
+func (r *ScaleREST) New() runtime.Object {
+	return &autoscaling.Scale{}
+}
+
+// Get retrieves object from Scale storage.
+func (r *ScaleREST) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	obj, err := r.store.Get(ctx, name, options)
+	if err != nil {
+		return nil, err
```

Comment:
we probably want to do the error remapping as we do for example here, no? https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/apps/deployment/storage/storage.go#L290

Reply:
The approach to errors in apps APIs differ. Should we unify them all?

```go
??? // DaemonSets
err // StatefulSets 
errors.NewNotFound(apps.Resource("deployments/scale"), name)
errors.NewNotFound(apps.Resource("replicasets/scale"), name)
errors.NewNotFound(autoscaling.Resource("replicationcontrollers/scale"), name)
```

Reply:
I think that is a good idea, but probably needs a separate PR for the StatefulSet?

Reply:
I'll do a separate one.

What do you think about replicationcontrollers/scale?

Reply:
Sorry, I have overlooked this

```
	default:
		return autoscalingv1.SchemeGroupVersion.WithKind("Scale")
```

Since we are returning `autoscaling/v1` object in all of these places it makes sense to me to use `errors.NewNotFound(autoscaling.Resource("$x/scale"), name)` everywhere. But somebody should probably double check me on this

Reply:
To my view this is an interpretation question. We have three possible cases represented above. 

1. If the err is nil, the API responds there is no parent resource. But it does not respond for the /scale subresource specifically.


$ kubectl get --raw /apis/apps/v1/namespaces/default/statefulsets/no-such-name/scale -v5
I0328 13:15:00.403257 1172604 helpers.go:216] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "statefulsets.apps \"no-such-name\" not found",
  "reason": "NotFound",
  "details": {
    "name": "no-such-name",
    "group": "apps",
    "kind": "statefulsets"
  },
  "code": 404
}]
$ kubectl get --raw /apis/apps/v1/namespaces/default/deployments/no-such-name/scale -v5
I0328 13:14:45.771847 1170382 helpers.go:216] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "deployments/scale.apps \"no-such-name\" not found",
  "reason": "NotFound",
  "details": {
    "name": "no-such-name",
    "group": "apps",
    "kind": "deployments/scale"
  },
  "code": 404
}]

Error from server (NotFound): statefulsets.apps "no-such-name" not found

$ kubectl get --raw /apis/apps/v1/namespaces/default/replicasets/no-such-name/scale -v5
I0328 13:15:31.863413 1176424 helpers.go:216] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "replicasets/scale.apps \"no-such-name\" not found",
  "reason": "NotFound",
  "details": {
    "name": "no-such-name",
    "group": "apps",
    "kind": "replicasets/scale"
  },
  "code": 404
}]

Reply:
I believe there must be rule of thumb. To my view this is an interpretation question. We have three possible cases. 

1. If the err is left unwrapped, the API says there is no parent resource. But it does not respond for the /scale subresource specifically.

<details>
  <summary>API call</summary>

```
$ kubectl get --raw /apis/apps/v1/namespaces/default/statefulsets/no-such-name/scale -v5
I0328 13:15:00.403257 1172604 helpers.go:216] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "statefulsets.apps \"no-such-name\" not found",
  "reason": "NotFound",
  "details": {
    "name": "no-such-name",
    "group": "apps",
    "kind": "statefulsets"
  },
  "code": 404
}]
```
</details>

2. If we map to `apps.Resource("$x/scale")`, the API responds with "$x/scale" being in "apps" group. But scale is from another group initially.

<details>
  <summary>API call</summary>

```
$ kubectl get --raw /apis/apps/v1/namespaces/default/deployments/no-such-name/scale -v5
I0328 13:14:45.771847 1170382 helpers.go:216] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "deployments/scale.apps \"no-such-name\" not found",
  "reason": "NotFound",
  "details": {
    "name": "no-such-name",
    "group": "apps",
    "kind": "deployments/scale"
  },
  "code": 404
}]
```

</details>

3. If we map to `autoscaling.Resource("$x/scale")`, the API responds with "$x/scale" being in "autoscaling" group. Bout routing-wise it is not.

Reply:
+1 for enumating the cases

can somebody from api help here? @lavalamp @liggitt

Reply:
I think `errors.NewNotFound(apps.Resource("$x"), name)` is correct. The thing that doesn't exist is the main resource. The subresource is just window dressing.

The only other alternative is to implement a e.g. SubResource(mainResource, subResource) function to express in details that yes, we know you didn't ask for the main resource. But I think this is overkill.

Reply:
> The thing that doesn't exist is the main resource.

Does that mean we can skip error mapping at all? All scale stores in apps are wrappers around their parent resources. 

BTW, now all mapped scale errors in Get hide internal store problems with NotFound.

https://github.com/kubernetes/kubernetes/blob/f14d1c9b1ef2b3b332d6b83d10da27fe3855acad/staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go#L42-L53

Reply:
We should do the thing that produces the right errors and doesn't e.g. mislabel other errors. (sorry no concrete suggestions)